### ISSUE TITLE
[ENH] MultiplexTransformer Discussion

### ISSUE THREAD

**Issue by miraep8** [alan-turing-institute/sktime#2670]
During the community meetup today @aiwalter and @indinewton suggested it would be useful to create this issue to discuss the implementation of a  `MultiplexTransformer`.  After getting a chance to check out `OptionalPassthrough` I think we might not need build this thing at all (I agree it could be a bad idea to have multiple areas of the code base with such overlapping/similar functionality).  But - it did leave me with some questions/ideas about some potential use cases that might be worth expanding the current model to support: 

#### Some Background:

- The original idea  to make a `MultiplexTransformer` came from a suggestion in #2459 that it would be nice to create a `MultiplexTransformer` to complement the `MultiplexForecaster`.
- As of #2612 we have implemented a `_DelegatedTransformer` to facilitate implementation of `MultiplexTransformer` in a similar way to how `MultiplexForecaster` has been written. 
- As was mentioned in the community meeting however - it seems like the `OptionalPassthrough` transformer can already do most of what we would want to do with the `MultiplexTransformer` though performing a grid search over its `passthrough` variable.  (A solution which I was not aware of, thank you for bringing it up!)

#### Some remaining questions/ideas I had:

-  **Question 1 - What if we want to use exactly 1 Transformer from a list of `n` transformers?**.  It seems to me with the current system if we did something like this (borrowed from the docstring example for `OptionalPassthrough`):

```
pipe = TransformedTargetForecaster(steps=[
    ("deseasonalizer", OptionalPassthrough(Deseasonalizer())),
    ("scaler", OptionalPassthrough(TabularToSeriesAdaptor(StandardScaler()))),
    ("forecaster", NaiveForecaster())])
# putting it all together in a grid search
cv = SlidingWindowSplitter(
    initial_window=60,
    window_length=24,
    start_with_window=True,
    step_length=48)
param_grid = {
    "deseasonalizer__passthrough" : [True, False],
    "scaler__transformer__transformer__with_mean": [True, False],
    "scaler__passthrough" : [True, False],
    "forecaster__strategy": ["drift", "mean", "last"]}
gscv = ForecastingGridSearchCV(
    forecaster=pipe,
    param_grid=param_grid,
    cv=cv,
    n_jobs=-1)
gscv_fitted = gscv.fit(load_airline())
```
Then we could actually end up with a situation where we either skip both `deseasonalizer` and `scaler` or include them both.  One might imagine it *could* be nice in some situations to specify that we want exactly one transformer from a list of transformers. Having a `MultiplexTransformer` is one way to achieve this (I didn't see a good way to do this with the current system), though I imagine it could also be supported by making some modifications to `param_grid` (ie could we specify pairs of variables simultaneously).  Personally I like the second idea, but I am not sure that it is the "best solution"

- **Question 2: What is the best way to support "multiplexing" over pairs of transformers/forecasters simultaneously?**  (eg, maybe some transformations are needed for certain forecasters).  The way that springs to mind would be to create a separate `ForecastingPipeline` or a `TransformedTargetForecaster` for each pair, and then put all the pairs into a `MultiplexForecaster`?   Would it be worth building a cleaner work around? (a similar question/solution springs to mind in the case where we had a list of known transformations to use but we wanted to select the optimal ordering ie `trans_a` -> `trans_b` vs `trans_b` -> `trans_a`) 
 

#### What should be resolved:

- [ ] Identify any meaningful use cases that either a) something like a `MultiplexTransformer` could uniquely support or b) that are worth expanding the current `OptionalPassthrough` approach to support as well?
- [ ] Should we clean up (ie delete) the `DelegatedTransformer` code if we don't need it? 

Lastly - I wonder if there is some way we can more easily inform people of the existence of `OptionalPassthrough`?  (Clearly people know about it!!  But I had missed it (which may of course just be on me)!  Thoughts on a Documentation page which has all the various Pipeline/model selection components? (and related dunders?))

**Comment by fkiraly** [comment]
Tagging related issue: https://github.com/alan-turing-institute/sktime/issues/2603

**Comment by fkiraly** [comment]
> As was mentioned in the community meeting however - it seems like the OptionalPassthrough transformer can already do most of what we would want to do with the MultiplexTransformer though performing a grid search over its passthrough variable. (A solution which I was not aware of, thank you for bringing it up!)

I was aware! I faintly remember that I may have said, written, or thought that `OptionalPassthrough` is, content-wise, equivalent to a `MultiplexTransformer` with the transformer that is the parameter, and an `IdentityTransformer`. So, the `OptionalPassthrough` would be the special case, the other way round I think it does not work.

Although I´m not sure whether implementing it that way the best solution, it seems to me the two transformers - multiplexer, and `OptionalPassthrough` - are not obvious special cases of each other, interface-wise. It might still be good to have them implemented separately.

**Comment by fkiraly** [comment]
> it seems like the OptionalPassthrough transformer can already do most of what we would want to do with the MultiplexTransformer though performing a grid search over its passthrough variable. (A solution which I was not aware of, thank you for bringing it up!)

@miraep8, @aiwalter, can you explain how this would look like mathematically?

This does not seem correct to me. I do not think we can easily replace an instance of the `MultiplexTransformer` with `OptionalPassthrough`.

How would this work, formally? Can you give an example, let´s say, how we would replace a multiplexer over three transformers, A, B, C?

**Comment by fkiraly** [comment]
>  The way that springs to mind would be to create a separate ForecastingPipeline or a TransformedTargetForecaster for each pair, and then put all the pairs into a MultiplexForecaster?

That's how I would do it. I.e., `(A * B) | (C * D)`, where `A`, `C` are transformers, and `B`, `D` are forecasters. The dunder syntax would invoke the `MultiplexForecaster`, not the `MultiplexTransformer`, since the pipelines `A * B` and `C * D` are forecasters.

**Comment by fkiraly** [comment]
> Identify any meaningful use cases that either a) something like a MultiplexTransformer could uniquely support or b) that are worth expanding the current OptionalPassthrough approach to support as well?

I think that already the simplest case, i.e., `A | B` , cannot be captured equivalently by a composition that involves only `OptionalPassthrough`. 
(except when `A` or `B` is trivial)

**Comment by fkiraly** [comment]
> Should we clean up (ie delete) the `DelegatedTransformer` code if we don't need it?

Yes, but I doubt that we do not need it.

**Comment by fkiraly** [comment]
> Lastly - I wonder if there is some way we can more easily inform people of the existence of OptionalPassthrough? (Clearly people know about it!! But I had missed it (which may of course just be on me)! Thoughts on a Documentation page which has all the various Pipeline/model selection components? (and related dunders?))

Yes! A "Pipeline" notebook. `examples/05_pipelines.ipynb` where all the neat facts about pipeliens are collected in one place, and which has a list and examples with all the components and dunders!

**Comment by fkiraly** [comment]
... and all this was discussed in the 10min when I wasn't there?? You were busy, apparently...

**Comment by indinewton** [comment]
> How would this work, formally? Can you give an example, let´s say, how we would replace a multiplexer over three transformers, A, B, C?
Let's take an example (below is not the actual code but more of a pythonic way of explaining the current concepts of sktime)
```python
transformers = [ 
 ("log", OptionalPassthrough(LogTransformer())),
 ("box", OptionalPassthrough(BoxCoxTransformer())),
]
# you can bundle up even upto 20 transformers like this

# lets say you have a model list
model_list = [
 ("arima", AutoArima()),
 ("ets", AutoETS()),
]
# you can also add many more forecasters in this model_list
```

From here, we challenged that the outcome and functionality of MultiplexTransformer + MultiplexForecaster 
will be same as what is achieved below
```python
best_forecaster = []
for model in model_list:
    pipe = transformers + model
    gscv_model = ForecastingGridSearchCV(forecaster = pipe, cv=cv, param_grid=param_grid)
    gscv_model.fit(y, fh)
    best_forecaster.append(gscv_model.best_forecaster_)

multiplex = MultiplexForecasters(best_forecasters, fh)
multiplex.fit(y, fh)

multiplex.predict(fh)
```
So having a class of MultiplexTransformer allows user to save this 5 to 6 lines of code. Which is why we were wondering what is the additional value of this besides convenience for the user.

**Comment by fkiraly** [comment]
@indinewton, can you kindly be more precise, and specify two *estimators* that would be equivalent?

You are outlining a manual process and not an estimator, and it seems you are not entirely on one page with @miraep8 on how the `MultiplexForecaster` would be used (or what it formally does). You would be doing `ForecastingGridSearchCV(MultiplexForecaster)`, not the other way round.

**Comment by fkiraly** [comment]
> a similar question/solution springs to mind in the case where we had a list of known transformations to use but we wanted to select the optimal ordering ie trans_a -> trans_b vs trans_b -> trans_a

Hm, depends on how many, and which combinations we want to select over.
If only some combinations, I would still use the multiplexer.

If all combinations, one could create a `PermutationPipeline`, which takes as parameters the transformers in the chain, and an element of the permutation group over the set of transformers. Although that feels impractical for anything except a small number of elements, due to combinatorial explosion.

**Comment by indinewton** [comment]
> Question 1 - What if we want to use exactly 1 Transformer from a list of n transformers?. It seems to me with the current system if we did something like this (borrowed from the docstring example for OptionalPassthrough)

@miraep8 Actually this is currently not possible in sktime. Adding MultiplexTransformer will probably solve this, you are right.

> Then we could actually end up with a situation where we either skip both deseasonalizer and scaler or include them both. One might imagine it could be nice in some situations to specify that we want exactly one transformer from a list of transformers. Having a MultiplexTransformer is one way to achieve this...

Generally when we are building forecasting applications, we expect GridSearchCV to find answers for us. So if the answer would be _deaseasonlizer_ true and _scaler_ false, then that would be detected through GSCV. But what if user just wants to pick only one transformer and not more? That's not possible to control as of current sktime classes. So MultiplexTransformer will be helpful there.

Additionally after reading through rapidfire comments from @fkiraly, I also agree with below - 
> This claim does not seem correct to me. I do not think we can easily replace an instance of the MultiplexTransformer with OptionalPassthrough.

MultiplexTransformer will have more control over returned Transformers by controlling some parameter like return_n_best_transformers = 2. With OptionalPassthrough you can only benefit it when you wrap it in ForecastingGridSearchCV. So @fkiraly is right, actually. It seems it makes to add this in Sktime. So, the conceptual example I commented earlier, seems a bit out of context.

**Comment by indinewton** [comment]
> > a similar question/solution springs to mind in the case where we had a list of known transformations to use but we wanted to select the optimal ordering ie trans_a -> trans_b vs trans_b -> trans_a
> 
> Hm, depends on how many, and which combinations we want to select over. If only some combinations, I would still use the multiplexer.
> 
> If all combinations, one could create a `PermutationPipeline`, which takes as parameters the transformers in the chain, and an element of the permutation group over the set of transformers. Although that feels impractical for anything except a small number of elements, due to combinatorial explosion.

I also agree with what has been suggested here. Additionally transformers can get really messed up if orders are randomly permuted. For example apply MinMaxScaler before a transformer which might be doing divisions or logs (LogTransformer), because minmax guarantees presence of 0 in the transformed data. Additionally, in pipeline that we have built for customers, the order is carefully decided after reading through all the transformers in sktime and sklearn. An example of such a order is: HampelFilter() should be first -> immediately followed by Imputer() ->  other transformers -> (at the end) MinMaxScaler (Obviously passed to TabularToSeriesAdaptor)

**Comment by miraep8** [comment]
> can you explain how this would look like mathematically?
>
> This claim does not seem correct to me. I do not think we can easily replace an instance of the MultiplexTransformer with OptionalPassthrough.

@fkiraly - I 100% agree that the `OptionalPassthrough` cannot fully replace what `MultiplexTransformer` could do in its current state- **Question 1** above was meant to be one such highlight of this issue (which I think is the more general case of the `A` | `B` example that you mentioned). 

My point was really - should we either a) change how we handle `OptionalPassthrough` (which I think we could do) so that it can cover more of the general cases or b) just write the `MultiplexTransformer`.  

I do *think* we could make `OptionalPassthrough` behave more like `MultiplexTransformer` if we changed how we did the grid search. As I mentioned above briefly one could imagine specifying pairs of variables simultaneously (ie we do the grid search over (`deseasonalizer`, `scaler`) = [(True, False), (False, True)] (not that the syntax would necessarily look like that, but hopefully you get the gist).

Would `MultiplexTransformer` read better than that^.......  I think so. But I guess I was trying to strike a balance between creating the best solution and one that works within the code base we already have.

Perhaps in the end we can just have both and that's ok/my concerns that it will be confusing are unmerited. 

>  Yes! A "Pipeline" notebook. examples/05_pipelines.ipynb where all the neat facts about pipelines are collected in one place, and which has a list and examples with all the components and dunders!

Sounds like a good idea!  Haha, I actually went to look for it first without seeing your note it was not written yet and was disappointed. Might I suggest that when we get around to it we also write a tutorial or similar on the docs site??  It just seems like too important a feature in sktime for users to struggle to have to piece together on their own.

**Comment by miraep8** [comment]
(**Question 2** was meant to demonstrate another case that would be easier to solve with `MultiplexTransformer` than with `OptionalPassthrough` I realize now perhaps I should have called them cases not questions)

**Comment by miraep8** [comment]
However - I don't want to discount my suggestion about generalizing the `GridSearchCV` `param_grid` entirely as I think it could also unlock some interesting functionality/abilities to pair/have more control over what space we are searching over.  Which might be really powerful! :) But also might be really confusing/more of a mess than its worth.

Just to reiterate, the real question hidden in the block of text above is: 

- can we make `OptionalPassthrough` sufficiently general by making the `param_grid` argument of `GridSearchCV` more flexible (and is that worth it even if we could do that) or should we just write `MultiplexTransformer` and be ok with having both.

**Comment by fkiraly** [comment]
> Might I suggest that when we get around to it we also write a tutorial or similar on the docs site?? It just seems like too important a feature in sktime for users to struggle to have to piece together on their own.

Depends on whom you mean with "we", is that that the Royal "we"? 😁 

> I do think we could make `OptionalPassthrough` behave more like `MultiplexTransformer` if we changed how we did the grid search.

That does not sound like a good idea, since I feel it would make both the compositors and the grid search more specific to each other, hence less generic. This sounds like an instance of "conceptual diffusion", where you blur the line between multiple otherwise very clear concepts. You would essentially get objects that are "both" passthrough and multiplexer and a bit of grid search maybe. 

I agree with you that this would not be pretty, and I would advise to go with the design that is conceptually clearer (the original one). The "formal" design principle is separation of concerns: https://en.wikipedia.org/wiki/Separation_of_concerns

If possible, an object should do passthrough or multiplexing or grid search but not all of it, since these are concerns that are clearly conceptually separated on the formal level (in my opinion).

> entirely as I think it could also unlock some interesting functionality/abilities to pair/have more control over what space we are searching over

In the grid search, you would still be specifying the same grid, no matter whether you merge the inner components or not? Or am I not seeing this correctly?

> can we make `OptionalPassthrough` sufficiently general by making the `param_grid` argument of `GridSearchCV` more flexible (and is that worth it even if we could do that) or should we just write `MultiplexTransformer` and be ok with having both.

I think we can, but I also think it´s a bad idea, since it scores lower on "separation of concerns" and "conceptual clarity" imo.

**Comment by miraep8** [comment]
So - main takeaways - it seems we are all on board with going with implementing  `MultiplexTransformer` as originally planned.  Unless anyone voices additional concerns, I will proceed with the `MultiplexTransformer` implementation as planned 👍 

> > entirely as I think it could also unlock some interesting functionality/abilities to pair/have more control over what space we are searching over
>
> in the grid search, you would still be specifying the same grid, no matter whether you merge the inner components or not? Or am I not seeing this correctly?

If I understand your question - then no, the explored parameter space in the "merged" version would actually be smaller.   For example - in the example I gave above (ie the example I gave of  (deseasonalizer, scaler) = [(True, False), (False, True)])) the space of parameter combinations we are exploring (only looking at deseasonalizer and scaler) would be {(True, False), (False, True)} whereas in the original "unmerged" component we would evaluate {(True, True), (True, False), (False, True) (False, False)}.

In that^ example the artificial `|` we are effectively creating can of course be accomplished with the `MultiplexTransformer` instead, which as we have agreed is a "prettier solution".  The reason I said we might not want to discount it entirely though is that there might be other use cases where having the flexibility to manually reduce the space of combinations we are considering is that a) it might help reduce the computation costs of a potentially very complex pipeline with lots of tunable parameters and b) help us avoid exploring parameter combinations that we might expect to be a poor fit for the model.  We could even provide classes that could be passed to grid search cv to facilitate this smaller search space - eg the example above could be generalized to a "pick one of n" function which was provided with a list of transformers. 

But thinking on it more the idea might end up having more complexity than its worth! :)

**Comment by fkiraly** [comment]
> For example - in the example I gave above (ie the example I gave of (deseasonalizer, scaler) = [(True, False), (False, True)])) the space of parameter combinations we are exploring (only looking at deseasonalizer and scaler) would be {(True, False), (False, True)} whereas in the original "unmerged" component we would evaluate {(True, True), (True, False), (False, True) (False, False)}.

I don't see that, or I don't see why you are not comparing like with like.
Can we not specify both the Cartesian product grid [True, False] x [True, False] and the "exactly one" grid [(True, False), (False, True)] in both cases? It's just a matter of the param grid to pass, the parameter *space* seems to be the same?

> Unless anyone voices additional concerns

No objections.

**Comment by miraep8** [comment]
In my second example I was trying to suggest that we not explore full grid, but rather a reduced parameter space. (Which now that I say it like that does seem like it would be confusing for users to have in "gridSearchCV"  - but if we liked it would could make a more general "paramSearchCV" or something like that). 

Hence the emphasis is on the set of parameter combinations one could explore for the two parameters `deseasonalize` and `scaler`. 

In the original example, because grid search does enforce a Cartesian product we have: 
{(True, True), (True, False) (False, True), (False, False)}

In the second I was suggesting we given the user the option to define a *reduced* parameter space to explore: 
{(True, False) (False, True)}

Hence they could recreate the "effective or"

**Comment by miraep8** [comment]
I am not necessarily arguing we need to do things this way, and I don't think `GridSearchCV` would be the place to do them if we did want to.  I am also not sure how much the user would benefit from this degree of control.  Obviously we plan to have a better way to achieve what this example was trying to do (`MultiplexTransformer`), but I was wondering if there might be other conditions where such a capability might be useful.

**Comment by fkiraly** [comment]
> In the original example, because grid search does enforce a Cartesian product

Ah, I think that's where we are not on one page.
Grid search does not necessarily optimize over a Cartesian product, see for instance the example here:
https://scikit-learn.org/stable/modules/grid_search.html#exhaustive-grid-search
Arbitrary sums of grids (Carteisan products) are part of common formulations.

**Comment by miraep8** [comment]
Nice!  Good to know!
